magenta/music/musicxml_parser.py

```diff
@@ -0,0 +1,678 @@
+      if score_part_index >= len(self.score_parts):
+        score_part = ScorePart(self.state)
+        self.score_parts.append(score_part)
+
+      part = Part(child, self.score_parts[score_part_index], self.state)
+      self.parts.append(part)
+      score_part_index = score_part_index + 1
+      if self.state.time_position > self.total_time:
+        self.total_time = self.state.time_position
+
+  def gettimesignatures(self):
+    """Return a list of all the time signatures used in this score.
+
+    Does not support polymeter (i.e. assumes all parts have the same
+    time signature, such as Part 1 having a time signature of 6/8
+    while Part 2 has a simultaneous time signature of 2/4).
+
+    Ignores duplicate time signatures to prevent Magenta duplicate
+    time signature error. This happens when multiple parts have the
+    same time signature is used in multiple parts at the same time.
+
+    Example: If Part 1 has a time siganture of 4/4 and Part 2 also
+    has a time signature of 4/4, then only instance of 4/4 is sent
+    to Magenta.
+    """
+    time_signatures = []
+    for part in self.parts:
+      for measure in part.measures:
+        if measure.time_signature is not None:
+          if not measure.time_signature in time_signatures:
+            # Prevent duplicate time signatures
+            time_signatures.append(measure.time_signature)
+
+    return time_signatures
+
+  def getkeysignatures(self):
+    """Return a list of all the key signatures used in this score.
+
+    Support different key signatures in different parts (score in
+    written pitch).
+
+    Ignores duplicate key signatures to prevent Magenta duplicate key
+    signature error. This happens when multiple parts have the same
+    key signature at the same time.
+
+    Example: If the score is in written pitch and the
+    flute is written in the key of Bb major, the trombone will also be
+    written in the key of Bb major. However, the clarinet and trumpet
+    will be written in the key of C major because they are Bb transposing
+    instruments.
+
+    If no key signatures are found, create a default key signature of
+    C major.
+    """
+    key_signatures = []
+    for part in self.parts:
+      for measure in part.measures:
+        if measure.key_signature is not None:
+          if not measure.key_signature in key_signatures:
+            # Prevent duplicate key signatures
+            key_signatures.append(measure.key_signature)
+
+    if len(key_signatures) == 0:
+      # If there are no key signatures, add C major at the beginning
+      key_signature = KeySignature(self.state)
+      key_signature.time_position = 0
+      key_signatures.append(key_signature)
+
+    return key_signatures
+
+  def gettempos(self):
+    """Return a list of all tempos in this score.
+    If no tempos are found, create a default tempo of 120 qpm.
+    """
+    tempos = []
+    #for part in self.parts:
```

Comment:
Not sure what you want deleted here. I removed the comment, but I kept the initialization of the tempos[] array.